File: backend/security/sanitizer.py
import base64
import re
import unicodedata
from typing import NamedTuple


class SanitizeResult(NamedTuple):
    text: str
    injection_detected: bool


# Injection patterns to detect/replace
_INJECTION_PATTERNS = [
    re.compile(r"ignore\s+(all\s+)?(previous|prior|above|earlier)\s+instructions?", re.I),
    re.compile(r"system\s*:", re.I),
    re.compile(r"\bdo\s+not\s+follow\b", re.I),
    re.compile(r"\bjailbreak\b", re.I),
    re.compile(r"\bpretend\s+(you\s+are|to\s+be|that)\b", re.I),
    re.compile(r"\byou\s+are\s+now\b", re.I),
    re.compile(r"\bnew\s+instruction(s)?\b", re.I),
    re.compile(r"<\/?system>", re.I),
    re.compile(r"\[INST\]", re.I),
    re.compile(r"\bHuman\s*:\s*", re.I),
    re.compile(r"\bAssistant\s*:\s*", re.I),
    re.compile(r"override\s+(your\s+)?(instructions?|prompt|rules?)", re.I),
]

# High ratio of special characters signals encoded injection
_MAX_SPECIAL_CHAR_RATIO = 0.30


def _check_base64_injection(text: str) -> bool:
    """Detect injection attempts hidden in Base64-encoded tokens."""
    tokens = text.split()
    for token in tokens:
        if len(token) < 20:
            continue
        try:
            decoded = base64.b64decode(token + "==", validate=False).decode("utf-8", errors="ignore")
            for pattern in _INJECTION_PATTERNS:
                if pattern.search(decoded):
                    return True
        except Exception:
            continue
    return False


def _special_char_ratio(text: str) -> float:
    if not text:
        return 0.0
    special = sum(1 for c in text if not (c.isalnum() or c.isspace()))
    return special / len(text)


def sanitize(text: str) -> SanitizeResult:
    """
    Sanitize text through 4 layers of injection protection.
    Returns the cleaned text and whether injection was detected.
    """
    # Layer 1: Unicode normalization (neutralizes homoglyph attacks)
    text = unicodedata.normalize("NFKC", text)

    injection_detected = False

    # Layer 2: Regex pattern matching
    for pattern in _INJECTION_PATTERNS:
        if pattern.search(text):
            injection_detected = True
            text = pattern.sub("[FILTERED]", text)

    # Layer 3: Base64-encoded injection detection
    if _check_base64_injection(text):
        injection_detected = True

    # Layer 4: High special-char ratio signals encoded/obfuscated injection
    if _special_char_ratio(text) > _MAX_SPECIAL_CHAR_RATIO:
        injection_detected = True

    return SanitizeResult(text=text, injection_detected=injection_detected)

File: lib/sanitizer.ts
/**
 * Client-side and server-side input sanitizer for injection protection.
 * Mirrors the Python backend/security/sanitizer.py logic.
 */

const INJECTION_PATTERNS: RegExp[] = [
  /ignore\s+(all\s+)?(previous|prior|above|earlier)\s+instructions?/gi,
  /system\s*:/gi,
  /\bdo\s+not\s+follow\b/gi,
  /\bjailbreak\b/gi,
  /\bpretend\s+(you\s+are|to\s+be|that)\b/gi,
  /\byou\s+are\s+now\b/gi,
  /\bnew\s+instruction(s)?\b/gi,
  /<\/?system>/gi,
  /\[INST\]/gi,
  /\bHuman\s*:\s*/gi,
  /\bAssistant\s*:\s*/gi,
  /override\s+(your\s+)?(instructions?|prompt|rules?)/gi,
]

const MAX_SPECIAL_CHAR_RATIO = 0.3

function specialCharRatio(text: string): number {
  if (!text) return 0
  const special = [...text].filter((c) => !/[\w\s]/.test(c)).length
  return special / text.length
}

export interface SanitizeResult {
  text: string
  injectionDetected: boolean
}

export function sanitize(text: string): SanitizeResult {
  let injectionDetected = false
  let sanitized = text

  for (const pattern of INJECTION_PATTERNS) {
    if (pattern.test(sanitized)) {
      injectionDetected = true
      sanitized = sanitized.replace(pattern, '[FILTERED]')
    }
    pattern.lastIndex = 0
  }

  if (specialCharRatio(text) > MAX_SPECIAL_CHAR_RATIO) {
    injectionDetected = true
  }

  return { text: sanitized, injectionDetected }
}
